### PR TITLE
LPD-23074 - Allow mapping of array fields

### PR DESCRIPTION
**Pull request by juanjofgliferay**
## Story / Task
[LPD-11818](https://liferay.atlassian.net/browse/LPD-11818) / [LPD-23074](https://liferay.atlassian.net/browse/LPD-23074)

## Goal
The idea is to allow users to select complex array fields. Those array fields could include simple nested objects but, they can contain any unknown structure (nested arrays with nested objects, localized fields buried in a deep collection of objects, ...).

## Concerns
- We should release this feature under a FF to test as much as possible and cover unseen edge cases.
- We should think again how we handle `nestedFields` . Calculation in the backend side doesn't seem accurate to me.
- Are array fields sortable? Are children of arrays sortable? If not, we could check its sortability based on the presence of `[]` and `.` characters in the `fieldName` better than the field type.

## UI
[Screencast from 2024-04-17 16-07-09.webm](https://github.com/liferay-frontend/liferay-portal/assets/132653342/7f6099c4-d03a-4ce1-a8cd-76e356cd2b38)

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts [hunk 1]
```diff
@@ -49,47 +52,82 @@ function getValidFields({
...
+					}),
+					label: propertyKey,
+					name: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_PARENT_SUFFIX}`,
+					type: type ? type : 'object',
+				});
+
+				return;
+			}
+
+			if (
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts:104]: Enabling this type of extensions allow fine grain selection of common nested fields.
If it is a valid choice we could merge with the previous one (`if (propertyValue.$ref)`) as they handle items in the same way.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx [hunk 2]
```diff
@@ -92,6 +92,14 @@ const DefaultRenderer: React.FC<{
 		return <Wrapper options={options}>{value.label}</Wrapper>;
 	}
 
+	if (
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx:95]: Allow scalar arrays to render as a comma separated list

File: modules/apps/frontend-data-set/frontend-data-set-web/test/utils/resolveField.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+	},
+};
+
+describe('resolveField', () => {
+	it('is defined', () => {
+		expect(resolveField).toBeDefined();
+	});
+
+	describe('single property request', () => {
+		it('returns an object containing the name of the field requested, the item itself and the root property of the object requested', () => {
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/test/utils/resolveField.ts:171]: I'm not happy with the test descriptions :shrug: . At least, each `given` variable offers a clear example of how each type should behave.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts [hunk 4]
```diff
@@ -63,7 +63,21 @@ function getValidFields({
...
+						contextPath: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_DELIMITER}`,
+						schemaName: propertyValue.items.$ref.replace(
+							/^.*\//,
+							''
+						),
+						schemas,
+					}),
+					label: propertyKey,
+					name: `${contextPath}${propertyKey}${FDS_NESTED_FIELD_NAME_PARENT_SUFFIX}`,
+					type: type ? type : 'object',
```
Comment (dsanz) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts:98]: we do know `type === EFieldType.ARRAY` here so we should not need this check.

Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts:98]: Scalar arrays do not need this process. They go through [L124](https://github.com/liferay-frontend/liferay-portal/pull/4135/files#diff-8b819d07f50b442d5ec220fd20383c40b9713e088f455625482a2c6d9b8e9a3dR124), default processing.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx (hunk 2 above)
Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx:95]: This effectively makes DefaultRenderer an "array-aware" renderer. It's enough for now to cover the cases we want in the Users MVP initiative :tada:

**Comment by juanjofgliferay** [comment]
ci:test:sf

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx [hunk 5]
```diff
@@ -92,6 +92,14 @@ const DefaultRenderer: React.FC<{
...
+		typeof value === 'object' &&
+		Array.isArray(value) &&
+		typeof value[0] !== 'object'
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx:98]: What if `value[0]` is a `function`? :trollface:

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/resolveField.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	FDS_ARRAY_FIELD_NAME_DELIMITER,
+	FDS_ARRAY_FIELD_NAME_PARENT_SUFFIX,
+	FDS_NESTED_FIELD_NAME_DELIMITER,
+	FDS_NESTED_FIELD_NAME_PARENT_SUFFIX,
+} from '../constants';
+
+export const resolveField = function resolveField(
+	path: string | Array<string>,
+	item: any
+) {
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/resolveField.ts:16]: I am really not a fan of utils that return objects with unrelated values or a tuple. It adds a layer of awkward naming, as you need to somehow name the frankenstein structure. This should be a `get*` of _something_, but nothing really sounds right.

Would it maybe make sense to move the content of `resolveField` inline into `getLocalizedValue`? There is only one invocation of it `resolveField`, and it's not exported. Then, we can rename `getLocalizedValue` into something like `getRendererProps`, that returns `IRendererProps` objects. Tests can be on that util, and include localization aspect. What do you think?

We don't need to do something like this in this PR, but it's seems like a good moment, at least to consider it.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/resolveField.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		path.includes(FDS_ARRAY_FIELD_NAME_PARENT_SUFFIX) ||
+		path.includes(FDS_NESTED_FIELD_NAME_PARENT_SUFFIX)
+	) {
+		itemPath.pop();
+	}
+
+	const resolvedFieldname = itemPath[itemPath.length - 1];
+	let resolvedItem = undefined;
+
+	function flatten(field: Array<any>) {
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/resolveField.ts:47]: TS on functions is much more helpful if it includes returning types.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/resolveField.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+				}
+				else if (navigatedValue.constructor.name === 'Object') {
+					return navigatedValue[chunk];
+				}
+			}, item) || {};
+
+		if (Array.isArray(resolvedItem)) {
+			resolvedItem = flatten(resolvedItem);
+		}
+	}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/resolveField.ts:81]: I trust you this madness works fine :sweat_smile:  If you have some doubts, I can take a closer look.

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay few comments inline

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx (hunk 5 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx:98]: 🫤

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
+/**
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts:1]: @markocikos Refactor function to do only one thing. Add some examples trying to clarify what is happening :sweat_smile:

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts [hunk 10]
```diff
@@ -49,47 +52,80 @@ function getValidFields({
...
+
+				return;
+			}
+
+			if (
+				propertyValue.extensions &&
+				propertyValue.extensions['x-parent-map'] === 'properties'
+			) {
+				const parentSchemaName =
+					propertyKey.charAt(0).toUpperCase() + propertyKey.slice(1);
```
[the remark this answers is not in the document]
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts]: No need to capitalize, but check that the schema name matches the selected one. Usual examples in the UI were `status` or `creator` but can be anything (`AuditEvents`, `WhateverId`, ...)

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:sf

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx (hunk 5 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DefaultRenderer.tsx:98]: Should we check for `function` also?
I know that it is kind of _soft_ validation. The second item could be anything and will break anyway

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+ * item: {a: {b: {c: [1, 2, 3]}}}
+ * returns: [1, 2, 3]
+ *
+ * path: 'a.b*'
+ * item: {a: {b: {c: [1, 2, 3]}}}
+ * returns: {c: [1, 2, 3]}
+ *
+ * path: 'a[].b.c'
+ * item: {a: [{b: {c: 1, d: 3}}, {b: {c: 2, d: 3}}, {b: {c: 3, d: 3}}]}
+ * returns: [{c:1}, {c: 2}, {c: 3}]
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts:28]: Shouldn't return here be `[1, 2, 3]`? The expectation is type of array items that matches type of the leaf, `c`, and that is a number.

Btw, love the examples docs! :heart:

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts (hunk 9 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts:1]: Much better, now with single returns! :clap:

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay looks good, just a behavior question inline

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts (hunk 11 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts:28]: Returning `[1, 2, 3]` will be part of [`getLocalizedValue`](https://github.com/liferay-frontend/liferay-portal/pull/4135/files#diff-8fb8b733d3441cdfc6070de3afc7eab272f73cd50ff30d81388045401c9d571cR147) . Consider that the value can contain strings to be translated. 

Related test -> https://github.com/liferay-frontend/liferay-portal/pull/4135/files#diff-49ba7db4dc49e6466b818786bdabaff962cb9d0df69c1c281e73aa99d5180c87R233

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/getItemField.ts:28]: OK, makes sense :+1:

**Review by markocikos — APPROVED** [review]
@juanjofgliferay LGTM! :sparkles:

**Comment by juanjofgliferay** [comment]
ci:forward

**Comment by juanjofgliferay** [comment]
ci:test:relevant

**Comment by juanjofgliferay** [comment]
ci:test:sf

**Comment by juanjofgliferay** [comment]
ci:test:relevant
